Description: Allow published ports to be reached across isolated bridges
Bug: https://github.com/containers/netavark/issues/709
Origin: backport, https://github.com/containers/netavark/pull/1483
Backported-by: Albert Yau <albert@xddd.org>
Forwarded: not-needed
Last-Update: 2026-10-27

--- a/src/firewall/nft.rs
+++ b/src/firewall/nft.rs
@@ -32,6 +32,8 @@
 pub(crate) const MAX_HASH_SIZE: usize = 13;
 
+const CT_MARK_NAME: Cow<'_, str> = Cow::Borrowed("mark");
 const MASK: u32 = 0x2000;
+const DNAT_MASK: u32 = 0x1000;
 
 const MULTICAST_NET_V4: &str = "224.0.0.0/4";
 const MULTICAST_NET_V6: &str = "ff00::/8";
@@ -252,6 +254,62 @@
             ));
         }
 
+        let match_dnat_mask = |r: &schema::Rule| -> bool {
+            // Match on any rule that matches against DNAT_MASK
+            for statement in r.expr.deref() {
+                match statement {
+                    stmt::Statement::Match(m) => match &m.right {
+                        expr::Expression::Number(n) => {
+                            if *n == DNAT_MASK {
+                                return true;
+                            }
+                        }
+                        _ => continue,
+                    },
+                    _ => continue,
+                }
+            }
+            false
+        };
+        if get_matching_rules_in_chain(&existing_rules, FORWARDCHAIN, match_dnat_mask).is_empty() {
+            // Ensure we allow the DNAT packets before we consult the isolation chains
+            // Forward: ct mark & 0x1000 == 0x1000 accept
+            let mut accept_rule = make_rule(
+                Cow::Borrowed(FORWARDCHAIN),
+                Cow::Owned(vec![
+                    stmt::Statement::Match(stmt::Match {
+                        left: expr::Expression::BinaryOperation(Box::new(
+                            expr::BinaryOperation::AND(
+                                expr::Expression::Named(expr::NamedExpression::CT(expr::CT {
+                                    key: CT_MARK_NAME,
+                                    family: None,
+                                    dir: None,
+                                })),
+                                expr::Expression::Number(DNAT_MASK),
+                            ),
+                        )),
+                        right: expr::Expression::Number(DNAT_MASK),
+                        op: stmt::Operator::EQ,
+                    }),
+                    stmt::Statement::Accept(None),
+                ]),
+            );
+            let isolation_handle = get_matching_rules_in_chain(
+                &existing_rules,
+                FORWARDCHAIN,
+                get_rule_matcher_jump_to(ISOLATION1CHAIN.to_string()),
+            )
+            .first()
+            .and_then(|r| r.handle);
+            if let Some(handle) = isolation_handle {
+                if let schema::NfListObject::Rule(rule) = &mut accept_rule {
+                    rule.handle = Some(handle);
+                }
+                batch.add_cmd(schema::NfCmd::Insert(accept_rule));
+            } else {
+                batch.add(accept_rule);
+            }
+        }
         // Forward chain: jump NETAVARK-ISOLATION-1
         if get_matching_rules_in_chain(
             &existing_rules,
@@ -1039,6 +1097,19 @@
             right: expr::Expression::Number(host_port),
             op: stmt::Operator::EQ,
         }));
+        statements.push(stmt::Statement::Mangle(stmt::Mangle {
+            key: expr::Expression::Named(expr::NamedExpression::CT(expr::CT {
+                key: CT_MARK_NAME,
+                family: None,
+                dir: None,
+            })),
+            value: expr::Expression::BinaryOperation(Box::new(expr::BinaryOperation::OR(vec![
+                expr::Expression::Named(expr::NamedExpression::Meta(expr::Meta {
+                    key: expr::MetaKey::Mark,
+                })),
+                expr::Expression::Number(DNAT_MASK),
+            ]))),
+        }));
         statements.push(stmt::Statement::DNAT(Some(stmt::NAT {
             addr: Some(expr::Expression::String(Cow::Owned(ip.to_string()))),
             family: Some(if ip.is_ipv6() {
